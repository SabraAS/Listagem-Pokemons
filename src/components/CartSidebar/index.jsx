import PropTypes from 'prop-types';
import './index.scss';

const CartSidebar = ({ pokemons, onRemovePokemon, onConfirmTeam }) => {
  return (
    <aside className="cart-sidebar">
      <div className="cart-sidebar__background" />
      <h2 className="cart-sidebar__title">Sua equipe</h2>
      <div className="cart-sidebar__list">
        {pokemons.length > 0 ? (
          pokemons.map((pokemon) => (
            <div className="cart-sidebar__item" key={pokemon.id}>
              <div className="cart-sidebar__content">
                <div className="cart-sidebar__info">
                  <div className="cart-sidebar__x">X</div>
                  <div className="cart-sidebar__name">
                    <p>{pokemon.name}</p>
                  </div>
                </div>
                <button
                  aria-label={`Remover ${pokemon.name} da equipe`}
                  className="cart-sidebar__remove-button"
                  onClick={() => onRemovePokemon(pokemon.id)}
                >
                  X
                </button>
              </div>
              <p className="cart-sidebar__characteristic">
                {pokemon.characteristic}
              </p>
            </div>
          ))
        ) : (
          <p>Nenhum Pok√©mon adicionado</p>
        )}
      </div>
      <div className="cart-sidebar__footer">
        <button
          className="cart-sidebar__footer-button"
          disabled={pokemons.length === 0}
          onClick={onConfirmTeam}
        >
          Confirmar Equipe
        </button>
      </div>
    </aside>
  );
};

CartSidebar.propTypes = {
  pokemons: PropTypes.array.isRequired,
  onRemovePokemon: PropTypes.func.isRequired,
  onConfirmTeam: PropTypes.func.isRequired,
};

export default CartSidebar;
