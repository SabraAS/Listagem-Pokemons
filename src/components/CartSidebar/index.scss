.cart-sidebar {
  background: white;
  border-radius: 8px;
  box-shadow: -2px 0px 8px rgb(0 0 0 / 10%);
  height: 500px;
  padding: 24px;
  width: 400px;
  position: fixed;
  right: 64px;
  top: 64px;
  z-index: 1;
  display: flex;
  flex-direction: column;

  &__background {
    position: absolute;
    bottom: 0;
    top: 0;
    left: 0;
    right: 0;
    background-image: url('@/assets/cover-pokemon.webp');
    background-repeat: repeat;
    overflow: hidden;
    height: 100%;
    opacity: 0.1;
    z-index: -1;
    background-size: 100%;
  }

  // Animation para aparecer ap√≥s LCP
  @keyframes fadeInBackground {
    from {
      opacity: 0;
    }
    to {
      opacity: 0.1;
    }
  }

  &__title {
    font-size: 24px;
    font-weight: 600;
    margin-bottom: 24px;
  }

  &__list {
    display: flex;
    flex-direction: column;
    gap: 12px;
    overflow-y: auto;
  }

  &__item {
    border-bottom: 1px solid rgb(0 0 0 / 10%);
    padding: 12px 0;

    &:last-child {
      border-bottom: none;
    }
  }

  &__content {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
  }

  &__info {
    display: flex;
    flex-direction: row;
    align-items: center;
    gap: 8px;
  }

  &__x {
    color: #f4dc25;
    font-size: 20px;
    margin-right: 4px;
  }

  &__name {
    font-size: 16px;
    font-weight: 500;
  }

  &__characteristic {
    font-size: 14px;
    color: #4b5563;
    padding-left: 40px;
    margin-top: 4px;
  }

  &__remove-button {
    background-color: transparent;
    border: 1px solid #000;
    border-radius: 100%;
    color: #000;
    height: 24px;
    width: 24px;
    cursor: pointer;
    font-size: 14px;
    cursor: pointer;

    &:hover {
      border-color: #f4dc25;
      box-shadow: 0 4px 8px rgb(0 0 0 / 10%);
      transform: translateY(-1px);
    }
  }

  &__footer {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-top: auto;
  }

  &__footer-button {
    background-color: #f4dc25;
    border: none;
    border-radius: 8px;
    box-shadow: 0 2px 4px rgb(0 0 0 / 10%);
    color: #374151;
    cursor: pointer;
    font-size: 16px;
    font-weight: 600;
    padding: 12px 0px;
    transition: all 0.2s ease-in-out;
    width: 100%;

    &:hover:not(:disabled) {
      box-shadow: 0 4px 8px rgb(0 0 0 / 10%);
      transform: scale(1.01);
    }

    &:disabled {
      background: #e5e7eb;
      color: #9ca3af;
      cursor: not-allowed;
    }
  }
}
