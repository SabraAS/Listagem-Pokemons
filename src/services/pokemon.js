import axios from 'axios';

const BASE_URL = 'https://pokeapi.co/api/v2';

export const getPokemonList = async (limit = 40) => {
  try {
    const response = await axios.get(`${BASE_URL}/pokemon?limit=${limit}`);
    const pokemonList = response.data.results;

    return await Promise.all(
      pokemonList.map(async (pokemon) => {
        const pokemonData = await axios.get(pokemon.url);
        const description = await getPokemonCharacteristic(pokemonData.data.id);
        return {
          id: pokemonData.data.id,
          name: pokemonData.data.name,
          image: pokemonData.data.sprites?.other?.dream_world?.front_default,
          abilities: pokemonData.data.abilities,
          types: pokemonData.data.types,
          characteristic: description,
        };
      }),
    );
  } catch (error) {
    console.log('Erro ao buscar lista de Pokémon');
    throw error;
  }
};

export const getPokemonCharacteristic = async (id) => {
  try {
    const response = await axios.get(`${BASE_URL}/characteristic/${id}`);
    const data = response.data;
    const englishDescription = data.descriptions.find(
      (desc) => desc.language.name === 'en',
    );
    return englishDescription?.description || 'não possui';
  } catch (_e) {
    console.log('Erro ao buscar característica do Pokémon');
    return 'não possui';
  }
};
