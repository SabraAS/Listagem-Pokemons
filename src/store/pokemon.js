import { create } from 'zustand';

export const usePokemonStore = create((set) => ({
  pokemons: [],
  setPokemons: (pokemons) => set({ pokemons }),
  addPokemon: (pokemon) =>
    set((state) => ({ pokemons: [...state.pokemons, pokemon] })),
  removePokemon: (pokemonId) =>
    set((state) => ({
      pokemons: state.pokemons.filter((pokemon) => pokemon.id !== pokemonId),
    })),
  clearTeam: () => set({ pokemons: [] }),
}));
