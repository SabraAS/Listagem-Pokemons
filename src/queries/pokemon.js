import { useQuery } from '@tanstack/react-query';

import { getPokemonList } from '@/services/pokemon';

export const POKEMON_KEYS = {
  all: ['pokemons'],
  list: (limit) => [...POKEMON_KEYS.all, 'list', limit],
};

export const usePokemons = (limit = 20) => {
  return useQuery({
    queryKey: POKEMON_KEYS.list(limit),
    queryFn: () => getPokemonList(limit),
    retry: false,
    refetchOnWindowFocus: false,
    select: (response) => response,
  });
};
