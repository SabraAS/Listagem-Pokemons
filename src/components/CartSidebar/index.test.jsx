import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';

import CartSidebar from './index';

import { mockPokemons, mockSinglePokemon } from '@/test/mocks/pokemon';

describe('CartSidebar', () => {
  const defaultProps = {
    pokemons: mockPokemons,
    onRemovePokemon: vi.fn(),
    onConfirmTeam: vi.fn(),
  };

  const emptyProps = {
    pokemons: [],
    onRemovePokemon: vi.fn(),
    onConfirmTeam: vi.fn(),
  };

  describe('Rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<CartSidebar {...defaultProps} />);
      expect(container).toMatchSnapshot();
    });

    it('should render title correctly', () => {
      render(<CartSidebar {...defaultProps} />);
      const title = screen.getByTestId('cart-sidebar-title');
      expect(title).toBeInTheDocument();
      expect(title).toHaveClass('cart-sidebar__title');
    });

    it('should render empty message when no pokemons', () => {
      render(<CartSidebar {...emptyProps} />);
      const emptyMessage = screen.getByTestId('cart-sidebar-empty-message');
      expect(emptyMessage).toBeInTheDocument();
      expect(emptyMessage).toHaveTextContent('Nenhum Pokémon adicionado');
    });

    it('should render pokemon list when pokemons exist', () => {
      render(<CartSidebar {...defaultProps} />);
      const bulbasaur = screen.getByTestId('cart-sidebar-item-1');
      expect(bulbasaur).toHaveTextContent('bulbasaur');

      const charmander = screen.getByTestId('cart-sidebar-item-4');
      expect(charmander).toHaveTextContent('charmander');
    });

    it('should render pokemon characteristics', () => {
      render(<CartSidebar {...defaultProps} />);
      const characteristic1 = screen.getByTestId(
        'cart-sidebar-pokemon-characteristic-1',
      );
      expect(characteristic1).toHaveTextContent(
        'A strange seed was planted on its back at birth. The plant sprouts and grows with this Pokémon.',
      );
    });

    it('should render remove buttons for each pokemon', () => {
      render(<CartSidebar {...defaultProps} />);
      const removeButton1 = screen.getByTestId('cart-sidebar-remove-button-1');
      expect(removeButton1).toHaveTextContent('X');
      const removeButton2 = screen.getByTestId('cart-sidebar-remove-button-4');
      expect(removeButton2).toHaveTextContent('X');
    });

    it('should render pokemon names', () => {
      render(<CartSidebar {...defaultProps} />);
      const bulbasaurName = screen.getByTestId('cart-sidebar-pokemon-name-1');
      expect(bulbasaurName).toHaveTextContent('bulbasaur');
      const charmanderName = screen.getByTestId('cart-sidebar-pokemon-name-4');
      expect(charmanderName).toHaveTextContent('charmander');
    });

    it('should render confirm button enabled when pokemons exist', () => {
      render(<CartSidebar {...defaultProps} />);
      const confirmButton = screen.getByTestId('cart-sidebar-footer-button');
      expect(confirmButton).not.toBeDisabled();
    });

    it('should render confirm button disabled when no pokemons', () => {
      render(<CartSidebar {...emptyProps} />);
      const confirmButton = screen.getByTestId('cart-sidebar-footer-button');
      expect(confirmButton).toBeDisabled();
    });
  });

  describe('Interactions', () => {
    it('should call onRemovePokemon when remove button is clicked', () => {
      const mockOnRemovePokemon = vi.fn();
      render(
        <CartSidebar
          {...defaultProps}
          onRemovePokemon={mockOnRemovePokemon}
          pokemons={mockPokemons}
        />,
      );
      const removeButton = screen.getByTestId('cart-sidebar-remove-button-1');
      fireEvent.click(removeButton);
      expect(mockOnRemovePokemon).toHaveBeenCalledTimes(1);
      expect(mockOnRemovePokemon).toHaveBeenCalledWith(1);
    });

    it('should call onConfirmTeam when confirm button is clicked', () => {
      const mockOnConfirmTeam = vi.fn();
      render(
        <CartSidebar
          {...defaultProps}
          onConfirmTeam={mockOnConfirmTeam}
          pokemons={mockPokemons}
        />,
      );
      const confirmButton = screen.getByTestId('cart-sidebar-footer-button');
      fireEvent.click(confirmButton);
      expect(mockOnConfirmTeam).toHaveBeenCalledTimes(1);
    });

    it('should not call onConfirmTeam when button is disabled', () => {
      const mockOnConfirmTeam = vi.fn();
      render(
        <CartSidebar
          {...defaultProps}
          onConfirmTeam={mockOnConfirmTeam}
          pokemons={[]}
        />,
      );
      const confirmButton = screen.getByTestId('cart-sidebar-footer-button');
      fireEvent.click(confirmButton);
      expect(mockOnConfirmTeam).not.toHaveBeenCalled();
    });
  });

  describe('Edge cases', () => {
    it('should handle pokemon with empty characteristic', () => {
      const pokemonWithEmptyChar = {
        ...mockSinglePokemon,
        characteristic: '',
      };
      render(
        <CartSidebar {...defaultProps} pokemons={[pokemonWithEmptyChar]} />,
      );
      const characteristicElement = screen.getByTestId(
        'cart-sidebar-pokemon-characteristic-1',
      );
      expect(characteristicElement).toHaveTextContent(
        'Pokémon sem característica',
      );
    });

    it('should handle null pokemons prop without crashing', () => {
      render(
        <CartSidebar
          onConfirmTeam={vi.fn()}
          onRemovePokemon={vi.fn()}
          pokemons={null}
        />,
      );
      const emptyMessage = screen.getByTestId('cart-sidebar-empty-message');
      expect(emptyMessage).toBeInTheDocument();
      const confirmButton = screen.getByTestId('cart-sidebar-footer-button');
      expect(confirmButton).toBeDisabled();
    });

    it('should handle pokemon with null ID', () => {
      const pokemonWithNullId = {
        ...mockSinglePokemon,
        id: null,
        name: 'pokemonSemId',
      };
      const mockRemove = vi.fn();
      render(
        <CartSidebar
          onConfirmTeam={vi.fn()}
          onRemovePokemon={mockRemove}
          pokemons={[pokemonWithNullId]}
        />,
      );
      const itemElement = screen.getByText('pokemonSemId');
      expect(itemElement).toBeInTheDocument();
      const removeButton = screen.getByTestId(
        'cart-sidebar-remove-button-null',
      );
      fireEvent.click(removeButton);
      expect(mockRemove).toHaveBeenCalledWith(null);
    });

    it('should handle pokemon without name property', () => {
      const pokemonWithoutName = {
        ...mockPokemons[0],
        name: undefined,
        id: 999,
      };
      render(
        <CartSidebar
          onConfirmTeam={vi.fn()}
          onRemovePokemon={vi.fn()}
          pokemons={[pokemonWithoutName]}
        />,
      );
      const name = screen.getByTestId('cart-sidebar-pokemon-name-999');
      expect(name).toHaveTextContent('pokémon sem nome');
    });
  });

  describe('CSS structure', () => {
    it('should have correct main container class', () => {
      render(<CartSidebar {...defaultProps} />);
      const sidebar = screen.getByTestId('cart-sidebar');
      expect(sidebar).toBeInTheDocument();
      expect(sidebar).toHaveClass('cart-sidebar');
      expect(sidebar.tagName).toBe('ASIDE');
    });

    it('should have correct list structure', () => {
      render(<CartSidebar {...defaultProps} />);
      const listContainer = screen.getByTestId('cart-sidebar-list');
      expect(listContainer).toBeInTheDocument();
      expect(listContainer).toHaveClass('cart-sidebar__list');
    });

    it('should have correct title class', () => {
      render(<CartSidebar {...defaultProps} />);
      const title = screen.getByTestId('cart-sidebar-title');
      expect(title).toHaveClass('cart-sidebar__title');
    });

    it('should have correct item structure and classes', () => {
      render(<CartSidebar {...defaultProps} />);
      const item = screen.getByTestId('cart-sidebar-item-1');
      expect(item).toBeInTheDocument();
      expect(item).toHaveClass('cart-sidebar__item');

      const content = screen.getByTestId('cart-sidebar-content-1');
      expect(content).toBeInTheDocument();
      expect(content).toHaveClass('cart-sidebar__content');

      const name = screen.getByTestId('cart-sidebar-pokemon-name-1');
      expect(name).toBeInTheDocument();
      expect(name).toHaveClass('cart-sidebar__name');

      const characteristic = screen.getByTestId(
        'cart-sidebar-pokemon-characteristic-1',
      );
      expect(characteristic).toHaveClass('cart-sidebar__characteristic');
    });

    it('should have correct footer structure', () => {
      render(<CartSidebar {...defaultProps} />);
      const footer = screen.getByTestId('cart-sidebar-footer');
      expect(footer).toBeInTheDocument();
      expect(footer).toHaveClass('cart-sidebar__footer');

      const button = screen.getByTestId('cart-sidebar-footer-button');
      expect(button).toHaveClass('cart-sidebar__footer-button');
    });

    it('should have correct remove button class', () => {
      render(<CartSidebar {...defaultProps} />);
      const removeButton = screen.getByTestId('cart-sidebar-remove-button-1');
      expect(removeButton).toHaveClass('cart-sidebar__remove-button');
    });

    it('should have correct background element', () => {
      render(<CartSidebar {...defaultProps} />);
      const background = screen.getByTestId('cart-sidebar-background');
      expect(background).toBeInTheDocument();
      expect(background).toHaveClass('cart-sidebar__background');
    });
  });

  describe('Accessibility', () => {
    it('should have no accessibility violations with empty list', async () => {
      const { container } = render(<CartSidebar {...emptyProps} />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should have no accessibility violations with pokemon list', async () => {
      const { container } = render(<CartSidebar {...defaultProps} />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should use semantic aside element', () => {
      render(<CartSidebar {...defaultProps} />);
      const aside = screen.getByRole('complementary');
      expect(aside).toBeInTheDocument();
      expect(aside).toHaveClass('cart-sidebar');
    });

    it('should have proper heading structure', () => {
      render(<CartSidebar {...defaultProps} />);
      const heading = screen.getByTestId('cart-sidebar-title');
      expect(heading.tagName).toBe('H2');
    });

    it('should have proper button accessibility', () => {
      render(<CartSidebar {...defaultProps} />);
      const removeButton1 = screen.getByTestId('cart-sidebar-remove-button-1');
      const removeButton2 = screen.getByTestId('cart-sidebar-remove-button-4'); //id do charmander é 4
      expect(removeButton1).toHaveAttribute(
        'aria-label',
        'Remover bulbasaur da equipe',
      );
      expect(removeButton2).toHaveAttribute(
        'aria-label',
        'Remover charmander da equipe',
      );
    });

    it('should have keyboard accessible buttons', () => {
      render(<CartSidebar {...defaultProps} />);
      const confirmButton = screen.getByTestId('cart-sidebar-footer-button');
      confirmButton.focus();
      expect(document.activeElement).toBe(confirmButton);

      const removeButton = screen.getByTestId('cart-sidebar-remove-button-1');
      removeButton.focus();
      expect(document.activeElement).toBe(removeButton);
    });
  });
});
