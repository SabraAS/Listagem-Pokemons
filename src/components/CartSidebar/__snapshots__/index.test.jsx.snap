// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartSidebar > should match snapshot 1`] = `
<div>
  <aside
    class="cart-sidebar"
  >
    <div
      class="cart-sidebar__background"
    />
    <h2
      class="cart-sidebar__title"
    >
      Sua equipe
    </h2>
    <div
      class="cart-sidebar__list"
    >
      <div
        class="cart-sidebar__item"
      >
        <div
          class="cart-sidebar__content"
        >
          <div
            class="cart-sidebar__info"
          >
            <div
              class="cart-sidebar__x"
            >
              X
            </div>
            <div
              class="cart-sidebar__name"
            >
              <p>
                bulbasaur
              </p>
            </div>
          </div>
          <button
            aria-label="Remover bulbasaur da equipe"
            class="cart-sidebar__remove-button"
          >
            X
          </button>
        </div>
        <p
          class="cart-sidebar__characteristic"
        >
          Loves to eat
        </p>
      </div>
      <div
        class="cart-sidebar__item"
      >
        <div
          class="cart-sidebar__content"
        >
          <div
            class="cart-sidebar__info"
          >
            <div
              class="cart-sidebar__x"
            >
              X
            </div>
            <div
              class="cart-sidebar__name"
            >
              <p>
                charmander
              </p>
            </div>
          </div>
          <button
            aria-label="Remover charmander da equipe"
            class="cart-sidebar__remove-button"
          >
            X
          </button>
        </div>
        <p
          class="cart-sidebar__characteristic"
        >
          Highly curious
        </p>
      </div>
    </div>
    <div
      class="cart-sidebar__footer"
    >
      <button
        class="cart-sidebar__footer-button"
      >
        Confirmar Equipe
      </button>
    </div>
  </aside>
</div>
`;
